File: aether/go/live_test.go
// Live end-to-end test: a real headless Chrome session driven entirely through
// the pure-Aether engine from Go. The whole pipeline —
// Go -> cgo -> libselenium_core.so -> std.http.client -> chromedriver -> Chrome.
// Skips if chromedriver is absent.
package selenium

import (
	"net"
	"net/url"
	"os/exec"
	"strconv"
	"testing"
	"time"
)

const liveHTML = `<html><head><title>Aether Selenium</title></head>` +
	`<body><h1 id="hdr">Hello</h1>` +
	`<a href="#" id="lnk" class="nav">click me</a>` +
	`<input id="box" name="q"/></body></html>`

func freePort(t *testing.T) int {
	l, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	defer l.Close()
	return l.Addr().(*net.TCPAddr).Port
}

func waitUp(port int, d time.Duration) bool {
	deadline := time.Now().Add(d)
	for time.Now().Before(deadline) {
		c, err := net.DialTimeout("tcp", net.JoinHostPort("127.0.0.1", strconv.Itoa(port)), 500*time.Millisecond)
		if err == nil {
			c.Close()
			return true
		}
		time.Sleep(100 * time.Millisecond)
	}
	return false
}

func TestLiveChrome(t *testing.T) {
	driverBin, err := exec.LookPath("chromedriver")
	if err != nil {
		t.Skip("chromedriver not on PATH")
	}
	port := freePort(t)
	cmd := exec.Command(driverBin, "--port="+strconv.Itoa(port))
	cmd.Stdout = nil
	cmd.Stderr = nil
	if err := cmd.Start(); err != nil {
		t.Skipf("could not start chromedriver: %v", err)
	}
	defer func() {
		_ = cmd.Process.Kill()
		_, _ = cmd.Process.Wait()
	}()
	if !waitUp(port, 10*time.Second) {
		t.Skip("chromedriver did not come up")
	}

	drv, err := NewChrome("http://127.0.0.1:"+strconv.Itoa(port), Headless())
	if err != nil {
		t.Fatalf("NewChrome: %v", err)
	}
	defer drv.Quit()

	if drv.SessionID() == "" {
		t.Fatal("no session id after newSession")
	}
	t.Logf("session %s...", drv.SessionID()[:8])

	page := "data:text/html;charset=utf-8," + url.PathEscape(liveHTML)
	if err := drv.Get(page); err != nil {
		t.Fatalf("Get: %v", err)
	}

	title, err := drv.Title()
	if err != nil || title != "Aether Selenium" {
		t.Fatalf("Title = %q, err=%v", title, err)
	}

	hdr, err := drv.FindElement(ByID, "hdr")
	if err != nil {
		t.Fatalf("FindElement(ByID): %v", err)
	}
	if txt, _ := hdr.Text(); txt != "Hello" {
		t.Fatalf("hdr.Text = %q", txt)
	}

	lnk, err := drv.FindElement(ByClassName, "nav")
	if err != nil {
		t.Fatalf("FindElement(ByClassName): %v", err)
	}
	if tag, _ := lnk.TagName(); tag != "a" {
		t.Fatalf("lnk.TagName = %q", tag)
	}
	if err := lnk.Click(); err != nil {
		t.Fatalf("Click: %v", err)
	}

	box, err := drv.FindElement(ByCSS, "#box")
	if err != nil {
		t.Fatalf("FindElement(#box): %v", err)
	}
	if err := box.SendKeys("hello world"); err != nil {
		t.Fatalf("SendKeys: %v", err)
	}
	if v, _ := box.GetProperty("value"); v != "hello world" {
		t.Fatalf("box value = %v", v)
	}

	n, err := drv.ExecuteScript("return 40 + 2;")
	if err != nil {
		t.Fatalf("ExecuteScript: %v", err)
	}
	if f, ok := n.(float64); !ok || f != 42 {
		t.Fatalf("script returned %v", n)
	}

	// Negative path: a missing element yields a typed NoSuchElement error.
	if _, err := drv.FindElement(ByID, "does-not-exist"); !IsNoSuchElement(err) {
		t.Fatalf("want NoSuchElement, got %v", err)
	}

	t.Log("live Chrome smoke test green")
}

File: aether/ruby/spec/live_test.rb
# frozen_string_literal: true

# Live end-to-end test: a real headless Chrome session driven entirely through
# the pure-Aether engine from Ruby. The whole pipeline —
# Ruby -> Fiddle -> libselenium_core.so -> std.http.client -> chromedriver ->
# Chrome. Skips if chromedriver is absent.

require 'minitest/autorun'
require 'socket'
require 'cgi'

$LOAD_PATH.unshift File.expand_path('../lib', __dir__)
require 'selenium_core'

class LiveTest < Minitest::Test
  HTML = '<html><head><title>Aether Selenium</title></head>' \
         '<body><h1 id="hdr">Hello</h1>' \
         '<a href="#" id="lnk" class="nav">click me</a>' \
         '<input id="box" name="q"/></body></html>'

  def setup
    @driver_bin = which('chromedriver')
    skip 'chromedriver not on PATH' unless @driver_bin
    @port = free_port
    @cd = spawn(@driver_bin, "--port=#{@port}", out: File::NULL, err: File::NULL)
    skip 'chromedriver did not come up' unless wait_up(@port)
  end

  def teardown
    return unless @cd

    Process.kill('TERM', @cd)
    Process.wait(@cd)
  rescue Errno::ESRCH, Errno::ECHILD
    # already gone
  end

  def test_live_chrome
    driver = SeleniumCore::WebDriver.headless_chrome("http://127.0.0.1:#{@port}")
    begin
      refute_empty driver.session_id, 'no session id after newSession'

      # CGI.escape form-encodes spaces as "+"; a data: URL wants %20.
      page = 'data:text/html;charset=utf-8,' + CGI.escape(HTML).gsub('+', '%20')
      driver.get(page)
      assert_equal 'Aether Selenium', driver.title

      hdr = driver.find_element(SeleniumCore::By::ID, 'hdr')
      assert_equal 'Hello', hdr.text

      lnk = driver.find_element(SeleniumCore::By::CLASS_NAME, 'nav')
      assert_equal 'a', lnk.tag_name.downcase
      lnk.click

      box = driver.find_element(SeleniumCore::By::CSS_SELECTOR, '#box')
      box.send_keys('hello world')
      assert_equal 'hello world', box.property('value')

      assert_equal 42, driver.execute_script('return 40 + 2;')

      assert_raises(SeleniumCore::NoSuchElementError) do
        driver.find_element(SeleniumCore::By::ID, 'does-not-exist')
      end
    ensure
      driver.quit
    end
  end

  private

  def which(cmd)
    ENV['PATH'].split(File::PATH_SEPARATOR).each do |dir|
      path = File.join(dir, cmd)
      return path if File.executable?(path) && !File.directory?(path)
    end
    nil
  end

  def free_port
    s = TCPServer.new('127.0.0.1', 0)
    port = s.addr[1]
    s.close
    port
  end

  def wait_up(port, timeout = 10.0)
    deadline = Time.now + timeout
    while Time.now < deadline
      begin
        TCPSocket.new('127.0.0.1', port).close
        return true
      rescue Errno::ECONNREFUSED
        sleep 0.1
      end
    end
    false
  end
end
